docs/source/mar_specification.rst

```diff
@@ -0,0 +1,137 @@
+* **license.txt**: software license attached to the model, can be left blank if no license needed.
+
+Archive Format
+==============
+
+The MAR directory and its contents can be compressed into a zip file to constitute a single file package. When unzipped into a directory this file will reproduce the above directory structure, and should itself also be named after the model it contains.
+
+The Torchscript file format is also just a zip file with a specific structure. When creating such an archive with `save_net_with_metadata` a MAR-compliant Torchscript file can be created by including the contents of `metadata.json` as the `meta_values` argument of the function, and other files included as `more_extra_files` entries. These will be stored in a `extras` directory in the zip file and can be retrieved with `load_net_with_metadata` or with any other library/tool that can read zip data. In this format the `model.*` files are obviously not needed by `README.md` and `license.txt` can be added as more extra files.
+
+metadata.json File
+==================
+
+This file contains the metadata information relating to the model, including what the shape and format of inputs and outputs are, what the meaning of the outputs are, what type of model is present, and other information. The JSON structure is a dictionary containing a defined set of keys with additional user-specified keys. The mandatory keys are as follows:
+
+* **version**: version of the stored model.
+* **monai_version**: version of MONAI the MAR was generated on, later versions expected to work.
+* **pytorch_version**: version of Pytorch the MAR was generated on, later versions expected to work.
+* **numpy_version**: version of Numpy the MAR was generated on, later versions expected to work.
+* **optional_packages_version**: dictionary relating optional package names to their versions, these packages are not needed but are recommended to be isntalled with this stated minimum version.
+* **task**: plain-language description of what the model is meant to do.
+* **description**: longer form plain-language description of what the model is, what it does, etc.
+* **authorship**: state author(s) of the model.
+* **copyright**: state model copyright.
+* **network_data_format**: defines the format, shape, and meaning of inputs and outputs to the model, contains keys "inputs" and "outputs" relating named inputs/outputs to their format specifiers (defined below).
+
+Tensor format specifiers are used to define input and output tensors and their meanings, and must be a dictionary containing at least these keys:
+
+* **type**: what sort of data the tensor represents: "image", "label", etc.
+* **format**: what format of information is stored: "magnitude", "hounsfield", "kspace", "segmentation", "multiclass", etc.
+* **num_channels**: number of channels the tensor has, assumed channel dimension first.
+* **spatial_shape**: shape of the spatial dimensions of the form "[H]", "[H, W]", or "[H, W, D]"
```

Comment:
I am concerned that in some cases this will be too restrictive. I have a few models that do not have a fixed input shape, but may have restrictions on the input shape, e.g. `[32n, 32n, 2n]` for something that must be as multiple 32 down the first two axes and a multiple of 2 down the last.

Reply:
How about allowing "any size".  That might be the case while dealing with Digital Pathology Slides...

Reply:
Hi @GreySeaWolf , may I know what's the `any size` you mean? Can we use "*" to represent it as @ericspod described?

Thanks.

Reply:
Hi @ericspod ,

I got a question here: What's the expected data type of `spatial_shape` in the metadata? Seems you are using `str` in this example. We need to finalize the data type then check it with the schema. Otherwise, it's hard to check network input / output shape with scripts.
What do you think?

Thanks.

Reply:
Shapes are going to have to be defined as lists of either strings to accommodate "2n" or "*" as specifiers or integers for exact know sizes. For example, a 2D image that's 256x256 in the spatial dimension with any number of channels would be `["*", 256, 256]` by if the spatial dimensions had to be multiples `n` of a power `p` of 2 we would have `["*", "2**p*n", "2**p*n"]`. The expression syntax we would allow here may get a bit complicated and requires validation in separate steps.